snapshot_dbg_cli/breakpoint_utils.py

```diff
@@ -167,4 +179,71 @@ def normalize_breakpoint(bp, bpid=None):
 
   set_converted_timestamps(bp)
 
+  if bp['action'] == 'LOG':
+    if 'logLevel' not in bp:
+      bp['logLevel'] = 'INFO'
+
+    if 'logMessageFormat' not in bp:
+      bp['logMessageFormat'] = ''
+
+    bp['logMessageFormatString'] = merge_log_expressions(
+        bp['logMessageFormat'], bp.get('expressions', []))
+
   return bp
+
+
+def merge_log_expressions(log_format, expressions):
+  """Replaces each $N substring with the corresponding {expression}.
+
+  This function is intended for reconstructing an input expression string that
+  has been split using split_log_expressions.
+
+  Args:
+    log_format: A string containing 0 or more $N substrings, where N is any
+      valid index into the expressions array. Each such substring will be
+      replaced by '{expression}', where "expression" is expressions[N].
+    expressions: The expressions to substitute into the format string.
+  Returns:
+    The combined string.
+  """
+
+  def get_expression(m):
+    try:
+      index = int(m.group(0)[1:])
+      return f'{{{expressions[index]}}}'
+    except IndexError:
+      return m.group(0)
+
+  parts = log_format.split('$$')
+  return '$'.join(re.sub(r'\$\d+', get_expression, part) for part in parts)
+
+
+def get_logpoint_short_status(logpoint):
+  if not logpoint['isFinalState']:
+    return 'ACTIVE'
+
+  status_message = StatusMessage(logpoint)
+
+  # This would be unexpected, as logpoints expire, which is classified as an
+  # error.
+  if not status_message.is_error:
+    return 'COMPLETED'
+
+  refers_to = status_message.refers_to
+
+  if refers_to is None or len(refers_to) == 0:
+    refers_to = 'FAILED'
```

Comment:
Nit: it bothers me that line 240 states that refers_to always starts with BREAKPOINT_ yet this one does not.  Probably doesn't matter, but it is an inconsistency.